util/stream.go

```diff
@@ -135,61 +120,103 @@ func (m *MessageStream) outbound() {
 	}
 }
 
+type smallMessageError struct {
+	length int
+}
+
+func (e smallMessageError) Error() string {
+	return fmt.Sprintf("invalid message with length %d is received", e.length)
+}
+
+func newSmallMessageError(length int) *smallMessageError {
+	return &smallMessageError{
+		length: length,
+	}
+}
+
+func getMessageLength(reader *bufio.Reader) (int, error) {
+	msgLength, readErr := func(reader *bufio.Reader) (int, error) {
+		headerBytes, err := reader.Peek(4)
+		if err != nil {
+			return 0, err
+		}
+		messageType := int(headerBytes[1])
+		messageLength := int(binary.BigEndian.Uint16(headerBytes[2:4]))
+		if messageType != 4 {
+			return messageLength, nil
+		}
+		venderHeaderMessageBytes, err := reader.Peek(16)
+		if err != nil {
+			return 0, err
+		}
+		// Check if the message is Type_PacketIn2 or not.
+		experimenterType := binary.BigEndian.Uint32(venderHeaderMessageBytes[12:])
+		if experimenterType != 30 {
+			return messageLength, nil
+		}
+
+		// Process packetIn2 message experimenterType == openflow15.Type_PacketIn2
+		pktIn2MessageBytes, err := reader.Peek(20)
+		if err != nil {
+			return 0, err
+		}
+
+		// Check if the first property is NXPINT_PACKET in Type_PacketIn2 message or not.
+		pktProp := int(binary.BigEndian.Uint16(pktIn2MessageBytes[16:18]))
+		if pktProp != 0 {
+			return messageLength, nil
+		}
+
+		// Read packet length from property NXPINT_PACKET
+		pktLength := int(binary.BigEndian.Uint16(pktIn2MessageBytes[18:]))
+		if pktLength < messageLength {
+			return messageLength, nil
+		}
+		messageLength += 1 << 16
```

Comment:
I think this is the key part of the implementation: if we detect that there was a wrap around (`pktLength < messageLength`) then we add `2^16` to the message length. There should be a comment here.

Reply:
Added comment as "Restore actual length for PacketIn2: messageLength is 16-bit, add 2^16 when it overflows."